dataplane/saiserver/l2.go

```diff
@@ -39,25 +50,147 @@ func newL2mcGroup(mgr *attrmgr.AttrMgr, dataplane switchDataplaneAPI, s *grpc.Se
 	return mg
 }
 
-// TODO: Implement this.
-func (mg *l2mcGroup) CreateL2McGroup(context.Context, *saipb.CreateL2McGroupRequest) (*saipb.CreateL2McGroupResponse, error) {
+func (mg *l2mcGroup) CreateL2McGroup(ctx context.Context, req *saipb.CreateL2McGroupRequest) (*saipb.CreateL2McGroupResponse, error) {
 	id := mg.mgr.NextID()
+	// Update internal data
+	mg.groups[id] = map[uint64]*l2mcGroupMember{}
+	// Update attributes
+	l2mcAttrs := &saipb.L2McGroupAttribute{
+		L2McOutputCount: proto.Uint32(0),
+		L2McMemberList:  []uint64{},
+	}
+	mg.mgr.StoreAttributes(id, l2mcAttrs)
 	return &saipb.CreateL2McGroupResponse{Oid: id}, nil
 }
 
-// TODO: Implement this.
-func (mg *l2mcGroup) CreateL2McGroupMember(context.Context, *saipb.CreateL2McGroupMemberRequest) (*saipb.CreateL2McGroupMemberResponse, error) {
+func (mg *l2mcGroup) CreateL2McGroupMember(ctx context.Context, req *saipb.CreateL2McGroupMemberRequest) (*saipb.CreateL2McGroupMemberResponse, error) {
 	id := mg.mgr.NextID()
+	// Update table entry
+	r := fwdconfig.TableEntryAddRequest(mg.dataplane.ID(), L2MCGroupTable).AppendEntry(
+		fwdconfig.EntryDesc(fwdconfig.ExactEntry(fwdconfig.PacketFieldBytes(fwdpb.PacketFieldNum_PACKET_FIELD_NUM_L2MC_GROUP_ID).WithUint64(req.GetL2McGroupId())))).Build()
+	for _, p := range mg.groups {
+		r.Entries[0].Actions = append(r.Entries[0].Actions, &fwdpb.ActionDesc{
+			ActionType: fwdpb.ActionType_ACTION_TYPE_MIRROR,
+			Action: &fwdpb.ActionDesc_Mirror{
+				Mirror: &fwdpb.MirrorActionDesc{
+					PortId: &fwdpb.PortId{
+						ObjectId: &fwdpb.ObjectId{
+							Id: fmt.Sprint(p),
+						},
+					},
+				},
+			},
+		})
+	}
+	if _, err := mg.dataplane.TableEntryAdd(ctx, r); err != nil {
+		return nil, err
+	}
+	// Update internal data
+	if mg.groups[req.GetL2McGroupId()] == nil {
+		return nil, fmt.Errorf("L2MC group id %q not found", req.GetL2McGroupId())
+	}
+	if _, ok := mg.groups[req.GetL2McGroupId()][req.GetL2McOutputId()]; ok {
+		return nil, fmt.Errorf("found existing member %q", req.GetL2McOutputId())
+	}
+	mg.groups[req.GetL2McGroupId()][req.GetL2McOutputId()] = &l2mcGroupMember{
```

Comment:
shouldn't you update the group before modifying the table entry?

Reply:
Done